File: LRU/146_LRU_cache.py
class Node:
    def __init__(self, val: int = None, key: int = None):
        self.value = val
        self.next = None
        self.prev = None
        self.key = key


class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.head = Node()
        self.tail = Node()
        self.cache = {}
        self.head.next = self.tail
        self.tail.prev = self.head

    def add_to_head(self, node):
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node

    def remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1

        node = self.cache[key]
        self.remove(node)
        self.add_to_head(node)

        return node.value

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.value = value
            self.remove(node)
            self.add_to_head(node)

            return

        if len(self.cache) == self.capacity:
            last_used_node = self.tail.prev
            self.remove(last_used_node)
            del self.cache[last_used_node.key]

        node = Node(value, key)
        self.cache[key] = node
        self.add_to_head(node)

File: LRU/146_LRU_cache.go
type Node struct {
	Key   int
	Next  *Node
	Prev  *Node
	Value int
}

type LRUCache struct {
	Capacity int
	HashMap  map[int]*Node
	Head     *Node
	Tail     *Node
}

func Constructor(capacity int) LRUCache {
	hashMap := make(map[int]*Node, capacity)
	head := &Node{}
	tail := &Node{}
	// дувусвязный список
	head.Next = tail
	tail.Prev = head
	return LRUCache{
		Capacity: capacity,
		HashMap:  hashMap,
		Head:     head,
		Tail:     tail,
	}
}

func (this *LRUCache) Remove(node *Node) {
	node.Prev.Next = node.Next
	node.Next.Prev = node.Prev
}

func (this *LRUCache) AddToHead(node *Node) {
	node.Next = this.Head.Next
	node.Prev = this.Head
	this.Head.Next.Prev = node
	this.Head.Next = node
}

func (this *LRUCache) Get(key int) int {
	node, ok := this.HashMap[key]
	if !ok {
		return -1
	}

	this.Remove(node)
	this.AddToHead(node)
	return node.Value
}

func (this *LRUCache) Put(key int, value int) {
	node, ok := this.HashMap[key]
	if ok {
		this.Remove(node)
		this.AddToHead(node)
		node.Value = value
		return
	}

	// вторая проверка возможно лишняя если по условию capacity >= 1
	if len(this.HashMap) == this.Capacity && len(this.HashMap) > 0 {
		node := this.Tail.Prev
		this.Remove(node)
		delete(this.HashMap, node.Key)
	}

	newNode := &Node{Key: key, Value: value}
	this.HashMap[key] = newNode
	this.AddToHead(newNode)
}
